pkg/engine/rwbuffer.go

```diff
@@ -0,0 +1,80 @@
+package engine
+
+import (
+	"io"
+	"sync"
+)
+
+type BufWriter struct {
+	sync.RWMutex
+
+	data   []byte
+	cond   *sync.Cond
+	closed bool
+}
+
+type BufReader struct {
+	writer *BufWriter
+	offs   int
+}
+
+func NewBufWriter() *BufWriter {
+	writer := &BufWriter{}
+	writer.cond = sync.NewCond(writer)
+	return writer
+}
+
+func (w *BufWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+	data := make([]byte, len(p))
+	copy(data, p)
+	w.Lock()
+	w.data = append(w.data, data...)
+	w.Unlock()
+	w.cond.Broadcast()
+	return len(p), nil
+}
+
+func (w *BufWriter) Close() {
+	w.Lock()
+	w.closed = true
+	w.Unlock()
+	w.cond.Broadcast()
+}
+
+func NewBufReader(writer *BufWriter) io.Reader {
+	reader := &BufReader{
+		writer: writer,
+	}
+	return reader
+}
+
+func (r *BufReader) Read(p []byte) (int, error) {
+	r.writer.cond.L.Lock()
+	defer r.writer.cond.L.Unlock()
+
+	if r.offs < len(r.writer.data) {
+		size := min(len(r.writer.data)-r.offs, len(p))
+		newoffs := r.offs + size
+		copy(p, r.writer.data[r.offs:newoffs])
+		r.offs = newoffs
+		return size, nil
```

Comment:
the `copy` function already leverages the min(len(dst), len(src)): 
https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/builtin/builtin.go;l=139

so I think that this code can be simplified to: 
```suggestion
		
		n := copy(p, r.writer.data[r.offs:])
		r.off += n
		return n

```